File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/java/src/main/java/com/microsoft/microhack/catalog/service/CategorySlug.java
package com.microsoft.microhack.catalog.service;

import java.text.Normalizer;
import java.util.Locale;

/** Implements the frozen category-slug-v1 normalization algorithm. */
public final class CategorySlug {

    private CategorySlug() {
    }

    /** Normalizes a display name into a lowercase ASCII slug. */
    public static String normalize(String value) {
        if (value == null) {
            return "";
        }
        String decomposed = Normalizer.normalize(value.trim(), Normalizer.Form.NFKD)
                .toLowerCase(Locale.ROOT);
        StringBuilder slug = new StringBuilder();
        boolean separatorPending = false;
        for (int index = 0; index < decomposed.length(); ) {
            int codePoint = decomposed.codePointAt(index);
            index += Character.charCount(codePoint);
            int type = Character.getType(codePoint);
            if (type == Character.NON_SPACING_MARK
                    || type == Character.COMBINING_SPACING_MARK
                    || type == Character.ENCLOSING_MARK
                    || codePoint == '\''
                    || codePoint == '\u2019') {
                continue;
            }
            boolean asciiAlphanumeric = (codePoint >= 'a' && codePoint <= 'z')
                    || (codePoint >= '0' && codePoint <= '9');
            if (asciiAlphanumeric) {
                if (separatorPending && !slug.isEmpty()) {
                    slug.append('-');
                }
                slug.appendCodePoint(codePoint);
                separatorPending = false;
            } else {
                separatorPending = true;
            }
        }
        return slug.toString();
    }
}

File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/dotnet/src/LegoCatalog.App/Services/CategorySlug.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace LegoCatalog.App.Services;

/// <summary>
/// Implements the frozen category-slug-v1 normalization algorithm.
/// </summary>
public static partial class CategorySlug
{
    /// <summary>
    /// Normalizes a display name into its stable ASCII slug.
    /// </summary>
    public static string Normalize(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        var normalized = value.Trim().Normalize(NormalizationForm.FormKD).ToLowerInvariant();
        var builder = new StringBuilder(normalized.Length);
        foreach (var rune in normalized.EnumerateRunes())
        {
            var category = Rune.GetUnicodeCategory(rune);
            if (category is UnicodeCategory.NonSpacingMark
                or UnicodeCategory.SpacingCombiningMark
                or UnicodeCategory.EnclosingMark)
            {
                continue;
            }

            if (rune.Value is '\'' or '\u2019')
            {
                continue;
            }

            builder.Append(rune.ToString());
        }

        return NonAsciiAlphanumericRun()
            .Replace(builder.ToString(), "-")
            .Trim('-');
    }

    [GeneratedRegex("[^a-z0-9]+", RegexOptions.CultureInvariant)]
    private static partial Regex NonAsciiAlphanumericRun();
}
